### PR TITLE
[RFC] external command line with colors

### PR DESCRIPTION
**Pull request by bfredl**
Rebase of #6162 with highlight support added. Rather than group names, I think it is better to transmit highlight than name, because that opens up for combined highlights in the future (think a builtin parser + rainbow parens as plugin). 

Docs are still WIP. And needs more tests.

I also borrowed a function from #7082 for converting highlights.

### CODE REVIEW

**Comment by bfredl** [comment]
Update: actually transmit contents of blocks. b/w for now, but could support coloring later on. Also blocks could be retransmitted from start, in case parser changed highlight of earlier code because later code.

Todo: This should also be done on `redraw!` as well as retransmitting nested cmdline.

**Comment by bfredl** [comment]
Marking RFC. Testing by external UI implementors would be appreciated.

**Comment by bfredl** [comment]
Note to self: test handling of `inputsecret()`

**Comment by bfredl** [comment]
@justinmk and others, what do you think of splitting of "6. Remote Uis" into its own help file? As the UI protocol gets more rich it is arguably not just "serialization" anymore, rather its own collection of features.

**Comment by justinmk** [comment]
@bfredl I was hoping to put it in a new section in `api.txt`:

    UI Events                            *api-ui-events*

or something like that. Haven't done it yet because I was also hoping it could be generated by `gen_api_vimdoc.py`

**Comment by bfredl** [comment]
@justinmk I just pushed a WIP proposal. Please look at general structure, it is knowingly full of broken formatting and references...

Auto-generation would be useful, but I would have a shot at that in a separate PR. we can figure out the target structure first...

**Comment by bfredl** [comment]
I wouldn't like `api.txt` to grow arbitrarily big. But we could call it `api_events.txt` and stuff like buffer events there too. But I still want remote_ui to span several sections (one for each "widget")

File: runtime/doc/api.txt [hunk 1]
```diff
@@ -48,7 +48,7 @@ version.api_compatible	API is backwards-compatible with this level
 version.api_prerelease	Declares the current API level as unstable >
 			(version.api_prerelease && fn.since == version.api_level)
 functions		API function signatures
-ui_events		UI event signatures |rpc-remote-ui|
+ui_events		UI event signatures |remote-ui|
```
Comment (justinmk) [inline runtime/doc/api.txt]: `|ui|` is sufficient. Just like we claimed `|api|`, we can make "UI" a single idea when communicating with users. Details about "msgpack" and "remote" don't need to be called out in most contexts.

Comment (bfredl) [reply runtime/doc/api.txt]: Agree completely

File: runtime/doc/remote_ui.txt [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+Nvim screen on a grid of width × height cells. `options` must be
+a dictionary with these (optional) keys:
+	`rgb`			Controls what color format to use.
+				Set to true (default) to use 24-bit rgb
+				colors.
+				Set to false to use terminal color codes (at
+				most 256 different colors).
+	`ext_popupmenu`		Externalize the popupmenu. |ui-events-popupmenu|
+	`ext_tabline`		Externalize the tabline. |ui-events-tabline|
+	`ext_cmdline`		Externalize the cmdline. |ui-events-cmdline|
```
Comment (justinmk) [inline runtime/doc/remote_ui.txt]: I would prefer to keep the tags `ui-ext-foo`. "ext" is the token that means extensible/external widget.

File: runtime/doc/remote_ui.txt [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+	safely ignore it.
+
+
+["update_menu"]
+	The menu mappings changed.
+
+==============================================================================
+3. External Popupmenu Events				  *ui-events-popupmenu*
+
+Only sent if `ext_popupmenu` option is set in *remote-ui-options*
```
Comment (justinmk) [inline runtime/doc/remote_ui.txt]: `*ui-options*`. Let's move away from `remote-` prefix except where it's absolutely needed.

File: runtime/doc/remote_ui.txt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+Only sent if `ext_tabline` option is set in *remote-ui-options*
+
+["tabline_update", curtab, tabs]
+	Tabline was updated. UIs should present this data in a custom tabline
+	widget.
+	curtab:	  Current Tabpage
+	tabs:	  List of Dicts [{ "tab": Tabpage, "name": String }, ...]
+
+==============================================================================
+5. External Cmdline Events				    *ui-events-cmdline*
```
Comment (justinmk) [inline runtime/doc/remote_ui.txt]: Just `Cmdline Events`, don't need to call out "External".

**Comment by justinmk** [comment]
Structure looks fine. Doesn't need to be perfect initially.

File: runtime/doc/remote_ui.txt (hunk 2 above)
Comment (bfredl) [reply runtime/doc/remote_ui.txt]: the point was to make it in line with `ui-events-grid` and `ui-events-mode`. Or we could reverse them and have `ui-events-global` and then `ui-events-grid`, the "grid" is some sense a widget, just the most important one. When we get multi-grid support I would rather have a "set current grid" event (prefixed in every update batch) than duplicating every single grid event (alternatively, add it as a field to every event).

Comment (justinmk) [reply runtime/doc/remote_ui.txt]: Maybe neither is needed: `ui-grid`, `ui-mode`, `ui-tabline`, ...

In all other contexts the `ui-` prefix is omitted, this means it's an editor feature.

File: runtime/doc/ui.txt [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+		|------------------------ | dst_top    |
+		| src (moved down) and dst|            |
+		+-------------------------+ src_bot    |
+		| dst (still in SR)       |            |
+		|=========================| dst_bot    |
+		| (clipped below SR)      |            v
+		+-------------------------+
+<
+==============================================================================
+4. Popupmenu Events						 *ui-popupmenu*
```
Comment (justinmk) [inline runtime/doc/ui.txt]: could omit the section numbers, makes reordering easier.

File: src/nvim/ui.c [hunk 6]
```diff
@@ -166,6 +167,88 @@ void ui_event(char *name, Array args)
   }
 }
 
+/// @param[in] aep data to convert
+/// @param[out] out structure that will be sent to UIs
+HlAttrs attrentry2hlattrs(const attrentry_T *aep, bool use_rgb)
```
Comment (justinmk) [inline src/nvim/ui.c]: param doc is wrong

File: src/nvim/ui.c [hunk 7]
```diff
@@ -166,6 +167,88 @@ void ui_event(char *name, Array args)
...
+  if (attrs.italic) {
+    PUT(hl, "italic", BOOLEAN_OBJ(true));
+  }
+
+  if (attrs.reverse) {
+    PUT(hl, "reverse", BOOLEAN_OBJ(true));
+  }
+
+  // todo care, we send none; change order between none/invalid
+if (attrs.foreground != -1) {
```
Comment (justinmk) [inline src/nvim/ui.c]: comment seems odd. also, indent.

**Comment by justinmk** [comment]
Since @teto's PR was merged this needs a rebase. Should we put this in 0.2.1, or wait until 0.2.2? 0.2.2 will be a micro-release: https://github.com/neovim/neovim/milestone/17

File: src/nvim/api/private/helpers.c [hunk 8]
```diff
@@ -687,6 +703,22 @@ String cstr_to_string(const char *str)
...
 }
 
+/// Copies buffer to an allocated String.
+/// The resulting string is also NUL-terminated, to facilitate interoperating
+/// with code using C strings.
+///
+/// @param str the buffer to copy
+/// @return the resulting String, if the input string was NULL, an
+///         empty String is returned
+String buf_to_string(const char *buf, size_t len)
```
Comment (justinmk) [inline src/nvim/api/private/helpers.c]: maybe name it `cbuf_to_string` in analogy with the others.

File: src/nvim/eval.c [hunk 9]
```diff
@@ -11149,15 +11149,23 @@ void get_user_input(const typval_T *const argvars,
...
-    msg_start();
-    msg_clr_eos();
-    msg_puts_attr_len(prompt, p - prompt, echo_attr);
-    msg_didout = false;
-    msg_starthere();
+    if (p == NULL) {
+      p = prompt;
+    } else {
+      if (p == NULL) {
+        p = prompt;
```
Comment (justinmk) [inline src/nvim/eval.c]: looks duplicated

File: src/nvim/ex_getln.c [hunk 10]
```diff
@@ -2342,8 +2361,7 @@ enum { MAX_CB_ERRORS = 1 };
 /// Should use built-in command parser or user-specified one. Currently only the
 /// latter is supported.
 ///
-/// @param[in]  colored_ccline  Command-line to color.
-/// @param[out]  ret_ccline_colors  What should be colored. Also holds a cache:
+/// @param[in/out]  colored_ccline  Command-line to color. Also holds a cache:
 ///                                 if ->prompt_id and ->cmdbuff values happen
 ///                                 to be equal to those from colored_cmdline it
 ///                                 will just do nothing, assuming that ->colors
```
Comment (justinmk) [inline src/nvim/ex_getln.c]: Is the comment still correct? `colored_ccline` holds a cache?

Comment (ZyX-I) [reply src/nvim/ex_getln.c]: Also `in/out` is written `in,out`. Don’t know whether doxygen is recognizing one, other or both, but it is how it is written in the rest of the code base.

File: src/nvim/ui.c (hunk 6 above)
Comment (bfredl) [reply src/nvim/ui.c]: this is borrowed code from @teto:s PR that will disappear on rebase.

File: src/nvim/ui.c (hunk 7 above)
Comment (bfredl) [reply src/nvim/ui.c]: ditto

File: src/nvim/api/private/helpers.c (hunk 8 above)
Comment (bfredl) [reply src/nvim/api/private/helpers.c]: `str` and `char` both refers to "c types as opposed to their API representation". But the alliteration is nice I suppose....
Yes, seems `size` is more common.

File: src/nvim/ex_getln.c (hunk 10 above)
Comment (bfredl) [reply src/nvim/ex_getln.c]: @justinmk yes, if the cmdline is not changed, it reuses the old color .
@ZyX-I noted, I might have copied some non-doxygen docstring by accident.

**Comment by bfredl** [comment]
@justinmk hasn't every point release been going to be a "mirco-release"? Also I hoping to get the channels/channel_info/better URP sequence into 0.2.2... If this PR is litterally the only thing that stops you from releasing 0.2.1 _today_ then maybe postpone it, but otherwise, as I said a "behind flag" extension has significantly lower risk than, say, completely changing the way TUI feature dectection is done for everyone...

File: runtime/doc/ui.txt [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+
+["cmdline_pos", pos, level]
+	Change the cursor position in the cmdline.
+
+["cmdline_special_char", c, shift, level]
+	Display a special char in the cmdline at the cursor position. This is
+	typically used to indicate a pending state, like right after Ctrl-V.
+	if shift is true the text after the cursor should be shifted,
+	otherwise it should overshadow the char at the cursor.
+	Should be hidden at next cmdline_char
```
Comment (justinmk) [inline runtime/doc/ui.txt]: @bfredl should cmdline_char be cmdline_pos?

(Just let me know; I'll fix it during merge)

**Pull request merged**